direct/tpm2/policy_test.go

```diff
@@ -281,6 +282,189 @@ func TestPolicyOrUpdate(t *testing.T) {
+	}
+	for _, tt := range tests {
+		t.Run(tt.desc, func(t *testing.T) {
+			policyPCR := PolicyPCR{
+				PolicySession: sess.Handle(),
+				PcrDigest: tpm2b.Digest{
+					Buffer: tt.expectedDigest,
+				},
+				Pcrs: selection,
+			}
+
+			if err := policyPCR.Execute(thetpm); err != nil {
+				t.Fatalf("executing PolicyPCR: %v", err)
+			}
+
+			pgd := PolicyGetDigest{
+				PolicySession: sess.Handle(),
+			}
+			want, err := pgd.Execute(thetpm)
+			if err != nil {
+				t.Fatalf("executing PolicyGetDigest: %v", err)
+			}
+
+			// Use the policy helper to calculate the same policy
+			pol, err := NewPolicyCalculator(tpm.AlgSHA1)
+			if err != nil {
+				t.Fatalf("creating policy calculator: %v", err)
+			}
+			policyPCR.Update(pol)
+			got := pol.Hash()
+
+			if bytes.Equal(got.Digest, want.PolicyDigest.Buffer) != tt.expectedResult {
+				t.Errorf("policyPCR.Hash() = %x,\nwant %x", got.Digest, want.PolicyDigest.Buffer)
+			}
+		})
+	}
+}
+
+// Testing void input for PcrDigest, from TPM 2.0 Part 3, 23.7.
+func TestPolicyPCRVoidPcrDigest(t *testing.T) {
+	thetpm, err := simulator.OpenSimulator()
+	if err != nil {
+		t.Fatalf("could not connect to TPM simulator: %v", err)
+	}
+	defer thetpm.Close()
+
+	// Use a trial session to calculate this policy
+	sess, cleanup2, err := PolicySession(thetpm, tpm.AlgSHA1, 16, Trial())
+	if err != nil {
+		t.Fatalf("setting up policy session: %v", err)
+	}
+	defer func() {
+		t.Helper()
+		if err := cleanup2(); err != nil {
+			t.Errorf("cleaning up policy session: %v", err)
+		}
+	}()
+
+	PCRs, err := CreatePCRSelection([]int{0, 3, 7})
+	if err != nil {
+		t.Fatalf("Failed to create PCRSelection")
+	}
+	selection := tpml.PCRSelection{
+		PCRSelections: []tpms.PCRSelection{
+			{
+				Hash:      tpm.AlgSHA1,
+				PCRSelect: PCRs,
+			},
+		},
+	}
+
+	policyPCR := PolicyPCR{
+		PolicySession: sess.Handle(),
+		Pcrs:          selection,
+	}
+
+	if err := policyPCR.Execute(thetpm); err != nil {
+		t.Fatalf("executing PolicyPCR: %v", err)
+	}
+
+	pgd := PolicyGetDigest{
+		PolicySession: sess.Handle(),
+	}
+	want, err := pgd.Execute(thetpm)
+	if err != nil {
+		t.Fatalf("executing PolicyGetDigest: %v", err)
+	}
+
+	pcrRead := PCRRead{
+		PCRSelectionIn: selection,
+	}
+
+	pcrReadRsp, err := pcrRead.Execute(thetpm)
+	if err != nil {
+		t.Fatalf("Failed to read PCRSelection")
+	}
+
+	var expectedVal []byte
+	for _, digest := range pcrReadRsp.PCRValues.Digests {
+		expectedVal = append(expectedVal, digest.Buffer...)
+	}
+	t.Logf("expectedVal=%x", expectedVal)
+
+	// Hash algorithm must match the one in PolicySession.
+	expectedDigest := sha1.Sum(expectedVal)
+	t.Logf("expectedDigest=%x", expectedDigest)
+
+	// Use the policy helper to calculate the same policy
+	pol, err := NewPolicyCalculator(tpm.AlgSHA1)
+	if err != nil {
+		t.Fatalf("creating policy calculator: %v", err)
+	}
+	
+	pol.Update(tpm.CCPolicyPCR, policyPCR.Pcrs, expectedDigest)
```

Comment:
Instead of calling Update ourselves (which we don't expect callers to actually directly do) how about assembling a PolicyPCR instance with pcrDigest populated, and passing that one to the calculator. The one actually invoked in the trial session should have pcrDigest omitted.

Reply:
Resolved by creating a populated PolicyPCR (policyPCRPopulated) for the calculator.